### PR TITLE
API-E2E-HappyTest

### PR DESCRIPTION
**Pull request by shenry99**
### Issue tracking

- [F2F-XXX](https://govukverify.atlassian.net/browse/F2F-XXX)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: .DS_Store
Comment (CoralSmith) [inline .DS_Store:1]: Can we add this to the .gitignore please?

File: src/tests/api/claimedIdentity.test.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		);
+ 
+		assert200OK({ status: response.status, statusText: response.statusText });
+
+	});
+
+});
+
+
+// describe("UnHappy Path", () => {
```
Comment (CoralSmith) [inline src/tests/api/claimedIdentity.test.ts:33]: Are we a 'remove commented out code' repo @richa-misra?

File: src/tests/api/session.test.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+	beforeEach( async () =>{
+		const requestBody = {
+			target:"https://unb2uhs7rl.execute-api.eu-west-2.amazonaws.com/dev/",
+		};
+		stubResponse = await stubStartPost(requestBody);
+
+	});
+    
+	it("Given valid endpoint has been received", async () => {
```
Comment (CoralSmith) [inline src/tests/api/session.test.ts:15]: Are there permutations of JAR objects we accept?

our default one has shared claims, do we still get successfully create a session with no shared claims?

File: src/tests/api/session.test.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+		const response = await sessionPost(stubResponse.data.clientId, "");
+
+		assert401Unauthorized(response.status, response.statusText);
+
+		expect(response.data.message).toBe("Invalid request: Request failed to be decrypted");
+	});
+
+
+	it("400 response - Empty Client Id in request body", async () => {
```
Comment (CoralSmith) [inline src/tests/api/session.test.ts:42]: Can we also check that a request for another CRI fails? Or a request from an unknown client id? Would be good to know that we dont let _anyone_ call us.

File: src/tests/api/wellKnown.test.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { assert200OK, assert201Created, assert400BadRequest, assert401Unauthorized } from "../../utils/apiCommonAssertions";
+import { authorizationGet, claimedIdentityPost, sessionPost, stubStartPost, tokenPost, userInfoPost, wellKnownGet } from "../../utils/apiTestSteps";
+
+describe("Happy Path Get WellKnown", () => {
+
+	it("Check wellKnown", async () => {
+		const response = await wellKnownGet();
+         
+        
+		assert200OK({ status: response.status, statusText: response.statusText });
```
Comment (CoralSmith) [inline src/tests/api/wellKnown.test.ts:10]: Should we do some validation that we're exposing the key types we'll need to expose to work correctly?

File: src/tests/api/userinfo.test.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+		tokenResponse = await tokenPost(authResponse.data.authorizationCode.value, authResponse.data.redirect_uri);
+
+	});
+    
+	it("Generate jwt", async () => {
+		const response = await userInfoPost(tokenResponse.data.access_token);
+		console.log("Response XYZ", response);
+        
+		assert200OK({ status: response.status, statusText: response.statusText });
```
Comment (CoralSmith) [inline src/tests/api/userinfo.test.ts:32]: We should process the data in this response to make sure it matches the format we expect.

File: src/tests/api/session.test.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		};
+		stubResponse = await stubStartPost(requestBody);
+
+	});
+    
+	it("Given valid endpoint has been received", async () => {
+
+		const response = await sessionPost(stubResponse.data.clientId, stubResponse.data.request);
+
+		assert200OK({ status: response.status, statusText: response.statusText });
```
Comment (CoralSmith) [inline src/tests/api/session.test.ts:19]: We should check a session Id is returned.

File: src/tests/api/session.test.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		const requestBody = {
+			target:"https://unb2uhs7rl.execute-api.eu-west-2.amazonaws.com/dev/",
+		};
+		const stubResponse = await stubStartPost(requestBody);
+
+		const response = await sessionPost(stubResponse.data.clientId, "");
+
+		assert401Unauthorized(response.status, response.statusText);
+
+		expect(response.data.message).toBe("Invalid request: Request failed to be decrypted");
```
Comment (CoralSmith) [inline src/tests/api/session.test.ts:38]: If the rest body is empty the oauth spec says we should return the user back with the error `invalid_request`. Does the consumer of this endpoint know to map this message to that error?

File: src/utils/apiTestSteps.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
+import axios from "axios";
+const API_INSTANCE = axios.create({ baseURL:"https://unb2uhs7rl.execute-api.eu-west-2.amazonaws.com/dev" });
+
+export async function stubStartPost(body:any):Promise<any> {
+	try {
+
+		const path = "https://erveje5km8.execute-api.eu-west-2.amazonaws.com/dev/start";
```
Comment (CoralSmith) [inline src/utils/apiTestSteps.ts:7]: Could we parameterise these somehow? As it stands we wont be able to reuse this test suite on multiple environments.

**Review by CoralSmith — COMMENTED** [review]
The test steps are clean, but we'd need to be careful not to run these steps in parallel, because a few test steps depend on the functionality of others (because there's an inherent flow/order in how these endpoints can be called. 

I'm also worried what we've not added much variation into our names. Slim Pickens is hilarious, but we should get a good coverage over name variations, including ones with grammar like O'Leary or accents like Köhne.

File: src/tests/api/claimedIdentity.test.ts (hunk 1 above)
Comment (Z-Mohamoud) [reply src/tests/api/claimedIdentity.test.ts]: Commented out code has been removed

File: src/tests/api/session.test.ts (hunk 3 above)
Comment (Z-Mohamoud) [reply src/tests/api/session.test.ts]: Added a check for invalid client id to negative test scenarios

File: src/tests/api/session.test.ts (hunk 6 above)
Comment (Z-Mohamoud) [reply src/tests/api/session.test.ts]: Session id is validated in every happy path test that uses this value

File: src/utils/apiTestSteps.ts (hunk 8 above)
Comment (Z-Mohamoud) [reply src/utils/apiTestSteps.ts]: Request URL's have been added as environment variables, .env.example file added to repository and repo secrets added for use in GitHub workflows

File: src/.env.example [hunk 9]
```diff
@@ -0,0 +1,13 @@
+DEV_CRI_CIC_API_URL=
+DEV_IPV_STUB_URL=
+
+
```
Comment (CoralSmith) [inline src/.env.example:4]: Could we remove some new lines? 😁

File: src/tests/api/session.test.ts (hunk 7 above)
Comment (Z-Mohamoud) [reply src/tests/api/session.test.ts]: For invalid requests with an incorrect or blank Request Body and Client Id the only values that are returned that we can currently validate are status code and response message/data. We've validated that the response message contains  'Invalid request' but I'm not sure how we'd be able to validate that the consumer knows how to map this message to that error. @CoralSmith

**Review by richa-misra — APPROVED** [review]

**Review by vibhutikaushal — APPROVED** [review]

**Review by vibhutikaushal — DISMISSED** [review]

**Review by Z-Mohamoud — APPROVED** [review]

**Review by Z-Mohamoud — DISMISSED** [review]

**Review by vibhutikaushal — APPROVED** [review]

**Review by richa-misra — APPROVED** [review]

**Pull request closed without merging**